src/uucore/src/lib/features/fs.rs

```diff
@@ -125,92 +135,152 @@ fn resolve<P: AsRef<Path>>(original: P) -> IOResult<PathBuf> {
-            .join(original)
-    };
+        let current_dir = env::current_dir()?;
+        let canonical_dir = dunce::canonicalize(current_dir)?;
+        Ok(canonical_dir.join(path))
+    }
+}
 
-    let mut result = PathBuf::new();
+fn get_parts(path: &Path, resolve_parents: bool) -> (Option<&OsStr>, Vec<&OsStr>) {
     let mut parts = vec![];
 
     // Split path by directory separator; add prefix (Windows-only) and root
     // directory to final path buffer; add remaining parts to temporary
     // vector for canonicalization.
-    for part in original.components() {
+    let mut prefix = None;
+    for part in path.components() {
         match part {
-            Component::Prefix(_) | Component::RootDir => {
-                result.push(part.as_os_str());
-            }
+            Component::Prefix(_) | Component::RootDir => prefix = Some(part.as_os_str()),
             Component::CurDir => (),
             Component::ParentDir => {
-                parts.pop();
+                if resolve_parents {
+                    parts.pop();
+                } else {
+                    parts.push(part.as_os_str());
+                }
             }
             Component::Normal(_) => {
                 parts.push(part.as_os_str());
             }
         }
     }
+    (prefix, parts)
+}
 
-    // Resolve the symlinks where possible
-    if !parts.is_empty() {
-        for part in parts[..parts.len() - 1].iter() {
-            result.push(part);
-
-            if can_mode == CanonicalizeMode::None {
-                continue;
-            }
+fn path_buf_append<P>(buf: &mut PathBuf, paths: &[P])
+where
+    P: AsRef<Path>,
+{
+    for path in paths {
+        buf.push(path)
+    }
+}
 
-            match resolve(&result) {
-                Err(_) if can_mode == CanonicalizeMode::Missing => continue,
-                Err(e) => return Err(e),
-                Ok(path) => {
+fn resolve_all_links(
+    result: &mut PathBuf,
+    parts: Vec<&OsStr>,
+    mode: CanonicalizeMode,
+    resolve_parents: bool,
+) -> IOResult<()> {
+    match mode {
+        // Resolve no links.
+        CanonicalizeMode::None => {
+            path_buf_append(result, &parts);
+        }
+        // Resolve all links, ignoring all errors.
+        CanonicalizeMode::Missing => {
+            for part in parts {
+                result.push(part);
+                if let Ok(path) = resolve(&result, resolve_parents) {
                     result.pop();
                     result.push(path);
                 }
             }
         }
-
-        result.push(parts.last().unwrap());
-
-        if can_mode == CanonicalizeMode::None {
-            return Ok(result);
+        // Resolve all links, propagating all errors.
+        CanonicalizeMode::Existing => {
+            for part in parts {
+                result.push(part);
+                let path = resolve(&result, resolve_parents)?;
+                result.pop();
+                result.push(path);
```

Comment:
It took me a while to realize that `result.push(path)` will actually replace `result` with `path` since `path` will be an absolute path. Maybe
```suggestion
                *result = path;
```
could express this better? Also, `result.pop()` is unneeded. Alternatively `resolve` could take a `&mut PathBuf` and modify it in place, which would avoid an allocation.

I know this isn't really related to your change because that same pattern already existed previously.

Reply:
Makes sense, I'll update this block and the similar code in the other arms of the `match` statement as well.